### DISCUSSION TITLE
Generic Ideas - Thread

### DISCUSSION THREAD

**Discussion by negue** [negue/meme-box#135] [Ideas]
If you have an idea, feel free to post it here, and we can either all just "vote" by reacting with emote or discuss feature X here

**Comment by negue** [comment]
Text Media: - just using text or markdown

**Comment by negue** [comment]
In-App simple Image generator for example speech bubbles

**Reply by negue** [reply]
^ Maybe 2022.2 otherwise the version after that

**Comment by gigili** [comment]
Sending text information alongside media clips. When using twitch events would be good to be able to send a message such as X raided with N viewers along with the clip, or X cheered N bits so that it gets displayed on screen as well. But I guess that would be covered by Text Media suggestion.

**Reply by negue** [reply]
The `Text` would be a standalone media- what you are suggesting would be something like "clip accessories" or like attachements I want to have to that DEFINITELY :)

**Comment by gigili** [comment]
Crazy feature idea:

When you open the screens pages and click to edit a media assigned to that screen, you have the option to use the full screen or manual position. So what I was thinking is when you select the manual position you get a small canvas shown that allows you to import a screenshot (or if that info can be obtained from obs would be better) of your main scene and than have a small box that would represent your media shown over the scene image that you can move around and than the app would calculate what the position would be for that media instead of a user having to manually type position. 

I could maybe provide a sample screenshot that explains what I'm thinking here if it is needed.

**Reply by negue** [reply]
Maybe even with thos https://docs.deckdeckgo.com/components/drr/  to drag and drop clips

**Reply by negue** [reply]
To get the screenshot maybe obs websocket could be used? https://github.com/Palakis/obs-websocket/blob/4.x-current/docs/generated/protocol.md#takesourcescreenshot

**Comment by negue** [comment]
Example Data View, which combines the manage and mobile view and maybe even a target screen to show it live

**Comment by negue** [comment]
Would be anymore interested in a markdown "media" ? or someone wants to be able to do that in a custom html?

**Comment by jonathanbucci** [comment]
Suggestion: A stop media button in the mobile, preview, and on StreamDeck. That way if a clip is playing or a meme is being shown for a long time, the user has the ability to stop the video or force the OUT to start manually. This could also allow for looping videos and memes that show until candled.

**Reply by negue** [reply]
When working with looped medias I thought it should use this way:

- in the Visibility Setting: Use `Toggle` instead of `Play`
- first trigger shows it
- second trigger stops it

unless I broke it this should work - could you check again ?

**Comment by Glycode** [comment]
Shuffle Function for Triggers.

have an option for an event to have multiple actions.

example: trigger word !laugh triggers an event where one randomly chosen laughtrack starts playing.

**Reply by gigili** [reply]
I think this can be achieved with meta actions, where you assign multiple audio files to a same meta action which has the play random option, but @negue can you confirm this or have I mistaken it for something else ?

**Reply by negue** [reply]
Yes this would be the [`Meta`](https://github.com/negue/meme-box/blob/develop/tutorials/meta_media.md)   

My naming of features needs to be improved yes. This Meta one will be hopefully replaced by a visual scripting / logic component in the next version which then would hopefully make things easier

**Reply by Glycode** [reply]
thank you it worked :) haven't touched anything below video so I didnt even realise

**Reply by negue** [reply]
if you have any suggestions on how to name / layout / add infos at... (insert other things to improve :D ) please feel free to create a quick mock up or something ... any thing helps

making good / understandable UX is really hard :D - I can do the code but everything else is over me 😅

**Reply by Glycode** [reply]
Well now that I know of this function I have more things to try out.
A few things I have in mind tho:

1: While adding a new action (a video for example) I personally find it a little annoying to search for my files in a drop-down. I would prefer the windows explorer a lot more. Maybe even an option to add multiple files at once from a folder and assign a tag and screen.

2: An Option to assign a screen in the action menu while adding a file (I feel like it was there at some point but I cant seem to find it)

3: In the Screen Tab: While assigning media to a specific screen, maybe an option to select/deselect all. Also once you click on a file it gets assigned or unassigned. Maybe change that to marking all files first (with a checkbox for example) and then click on an assign/unassign button to assign/unassign.

4: to the meta action: maybe rename it to meta tagging or something

I'm still exploring but once I have any more suggestions I'll let you know. And I also wanted to say thanks, I really love this app, great job. 
If I weren't a poor student I'd donate something

**Reply by negue** [reply]
> While adding a new action (a video for example) I personally find it a little annoying to search for my files in a drop-down. I would prefer the windows explorer a lot more. 

yeah I think the current way the files are listed is just "read all files from the media folder", there is no upload so we can't use the browser "select files dialog" - would a text search inside the select help there? or would it be still too many clips

a visible browser for that hmmm don't think I'll have the time for that during the next releases - lots of features already planned

> Maybe even an option to add multiple files at once from a folder and assign a tag and screen.

when you are in the screen arrange view - there is a button `Assign Media`  there you can select and auto add multiple medias at once, would something like this help ?

> An Option to assign a screen in the action menu while adding a file (I feel like it was there at some point but I cant seem to find it)

It is there when adding a new media, during an edit it would make more confusion when you could add it to a new screen (since you can have multiples for different usecases) - there the workflow would be rather => select a screen -> assign a new media to it

How about a context menu on a media which has a "add to screen" option ?

> In the Screen Tab: While assigning media to a specific screen, maybe an option to select/deselect all. Also once you click on a file it gets assigned or unassigned. Maybe change that to marking all files first (with a checkbox for example) and then click on an assign/unassign button to assign/unassign.

I'd need a need a bit more information about that 😅  - also that as a different entry here - then its easier (like the new one you did)

> to the meta action: maybe rename it to meta tagging or something

yeah the meta one is just because ... I cant name things xD and since it can be used for "random trigger, trigger all at once - or trigger all in a delay"  this was before scripts and on the next release (not the one that is getting released soon) there will be some kind of visual scripting / logic - which would deprecate meta -  I could also name it "Special" but its also not really naming the function it has :D 

>  And I also wanted to say thanks, I really love this app, great job.

Thank you!! Its been now 1.5 years :D and its still not really there yet of what I have in mind - but working on it mostly alone just takes time xD

>  If I weren't a poor student I'd donate something

No worries!! - Ideas are good too ;) - also I don't even have any donate functionality 😅

**Reply by Glycode** [reply]
> would a text search inside the select help there? or would it be still too many clips

yeah it would be better than now at least. I've noticed it reads the files in order of their position in the media folder. So when I add new videos I create a new folder inside the media folder and name it so that it appears in the first position. something like 1_name, 11_name or aa_name idk. Just a little side note



> when you are in the screen arrange view - there is a button `Assign Media` there you can select and auto add multiple medias at once, would something like this help ?

I meant a step before that. Adding files to memebox in the first place. but I suppose thats too much rn.


> How about a context menu on a media which has a "add to screen" option ?

that sound like a good option. but not only "add to screen" but maybe also other things like volume for quick edits.
 

> > In the Screen Tab: While assigning media to a specific screen, maybe an option to select/deselect all. Also once you click on a file it gets assigned or unassigned. Maybe change that to marking all files first (with a checkbox for example) and then click on an assign/unassign button to assign/unassign.
> 
> I'd need a need a bit more information about that 😅 - also that as a different entry here - then its easier (like the new one you did)

Ok I'll make a new mockup later haha

**Comment by Glycode** [comment]
Just downloaded the current nightly. the screen option is now on the top of the add action screen which is nice. 
would it be possible to also add the tag option there? that way videos can be assigned to screen and tags without having to change to other settings. Also leave the play time on screen empty by default instead of 2500 ms

A trigger option for new followers would also be nice.

Maybe an option for the random meta action to not play the same file twice (before other files were played) by maybe creating a random order each session which resets after every file has been played. 

also how about a global volume option to change the volume of all videos at once. or an extra edit window where you can edit every video and once and then update. because if you edit just one video the whole page updates and you have to scroll again to find the next one.
I've made a mockup as an example

![memebox_edit](https://user-images.githubusercontent.com/97413783/155761630-678f137b-056e-4ed8-a57d-6c4ae3994391.png)
![memebox_action](https://user-images.githubusercontent.com/97413783/155761845-0d3bfece-ff8e-4db3-9d67-1a3b01370a99.png)

**Reply by negue** [reply]
> would it be possible to also add the tag option there?

hmmm maybe not a full input field - but maybe with an icon-button that opens another dialog which has all the tags there? OR ... hmm maybe a simple checklist that opens up I'll try what I can do - I don't even know if tags will stay after the next release (the meta overhaul I mentioned) 

> Also leave the play time on screen empty by default instead of 2500 ms

The time is needed for media that doesn't have a fixed length - for example audio / video staying empty cause then the duration of that file is used - for media like images / widgets / iframe - you (I or memebox) need a time to show it for this duration

It usually works like this:

Animate IN -> show media for duration (either the video/audio length or the one you set, 2500ms) -> Animate OUT -> hide

If you change the visibility from `Play`  ^  to `Always Visible` it'll be alway there or to `Toggle` when each time you trigger it - it'll be either shown or hidden without using the Duration Time


> A trigger option for new followers would also be nice.

I WISH if Twitch would just sent this event out but NOOO - they only send it for WebHooks (you'd need to have an online accessible server) just to get this info ... its stupid.  Soon there'll be "EventSub WS" which then I can use without having a server (I hope) - applied for the closed beta - no invitation sadly - soo I think we all need to wait for that ^^ 
But yes I want that too.

> Maybe an option for the random meta action to not play the same file twice (before other files were played) by maybe creating a random order each session which resets after every file has been played.

This will be definitely something once the new system replaced the current "Meta" one - then those options could be added to each task inside - you could do that using scripts though. if you have something this could be also merged as an example if you want :) 

> also how about a global volume option to change the volume of all videos at once. or an extra edit window where you can edit every video and once and then update. because if you edit just one video the whole page updates and you have to scroll again to find the next one.
I've made a mockup as an example

The full screen refresh shouldn't happen ... wonder why it broke again - its like each new month it breaks again even though it was fixed the month before 

I want to add a "pro" view - which has only the columns / data you want - and this would be something like you craeted in the mockup - like you choose I want to have the following data to be shown and be able to change it asap without going through multiple dialogs and stuff. 

YES to the idea - the only question is when :D

**Reply by Glycode** [reply]
> > would it be possible to also add the tag option there?
> 
> hmmm maybe not a full input field - but maybe with an icon-button that opens another dialog which has all the tags there? OR ... hmm maybe a simple checklist that opens up I'll try what I can do - I don't even know if tags will stay after the next release (the meta overhaul I mentioned)

I mean a drop-down or checklist would also work of course, there is no need for input if it only uses the previously made meta action tags.

> 
> > Also leave the play time on screen empty by default instead of 2500 ms
> 
> The time is needed for media that doesn't have a fixed length - for example audio / video staying empty cause then the duration of that file is used - for media like images / widgets / iframe - you (I or memebox) need a time to show it for this duration
> 
> It usually works like this:
> 
> Animate IN -> show media for duration (either the video/audio length or the one you set, 2500ms) -> Animate OUT -> hide
> 
> If you change the visibility from `Play` ^ to `Always Visible` it'll be alway there or to `Toggle` when each time you trigger it - it'll be either shown or hidden without using the Duration Time

okay I see
 
> > A trigger option for new followers would also be nice.
> 
> I WISH if Twitch would just sent this event out but NOOO - they only send it for WebHooks (you'd need to have an online accessible server) just to get this info ... its stupid. Soon there'll be "EventSub WS" which then I can use without having a server (I hope) - applied for the closed beta - no invitation sadly - soo I think we all need to wait for that ^^ But yes I want that too.

oof that sucks :/ 
well I'm using a workaround for that. The streamElements bot can register new follows so I have it write something specific every time I get a new follower which then triggers the memebox text trigger.


> > Maybe an option for the random meta action to not play the same file twice (before other files were played) by maybe creating a random order each session which resets after every file has been played.
> 
> This will be definitely something once the new system replaced the current "Meta" one - then those options could be added to each task inside - you could do that using scripts though. if you have something this could be also merged as an example if you want :)
> 
Awesome. Where can I learn how to create a script? What do you mean by if I have something? 

> > also how about a global volume option to change the volume of all videos at once. or an extra edit window where you can edit every video and once and then update. because if you edit just one video the whole page updates and you have to scroll again to find the next one.
> > I've made a mockup as an example
> 
> The full screen refresh shouldn't happen ... wonder why it broke again - its like each new month it breaks again even though it was fixed the month before
> 
> I want to add a "pro" view - which has only the columns / data you want - and this would be something like you craeted in the mockup - like you choose I want to have the following data to be shown and be able to change it asap without going through multiple dialogs and stuff.
> 
> YES to the idea - the only question is when :D

Can't wait haha

**Comment by Glycode** [comment]
Okay so I made a mockup of what I meant with the assign function.
Because if you create a new screen and want to assign every available media to it, it assigns them on every click which creates a delay for each media. This way you can select them first (either one by one by clicking on them or all at once) and then assign all at once. 

![Assign_Mockup](https://user-images.githubusercontent.com/97413783/155849840-36b20565-a87a-4b90-bb3a-e7f28d1f45ae.png)

**Comment by Glycode** [comment]
A queue function with different queues.
Triggers can then be applied to the different queues. 
Example:
Trigger 1 (in queue 1) gets activated and starts a video.
Trigger 2 (also in queue 1) gets activated too but can only start playing a video after the first video from trigger 1 is finished.
Trigger 3 (in queue 2) gets activated but can start playing because it's independent from the videos in queue 1.

**Reply by negue** [reply]
There are already queues, to be found under Media Settings -> other Settings
![other settings](https://user-images.githubusercontent.com/842273/157134733-3039e91f-c251-4c00-a185-8c2d8cf93ea3.png)

If thats different from what you thought let me know

**Reply by Glycode** [reply]
Oh my bad, no I think this is exactly that

**Comment by Glycode** [comment]
So while setting up memebox from the beginning I had to add lots of video. One helpful thing is the "import every media" option in the settings menu but their settings still have to be changed. 
Currently its the fastest for me to edit the config file.
An additional QoL update to that would be the option to batch import subfolders and maybe even batch edit those before import. Just like when adding actions where you can edit them (volumen, tags, queue, playtime, screen, etc...). 

And also maybe a general batch edit option where multiple videos can be edited at once (similar to the mockup I posted on Feb 25th)

**Reply by negue** [reply]
> Currently its the fastest for me to edit the config file.

I'm sorry but yeah so far thats the fastest way.

> An additional QoL update to that would be the option to batch import subfolders 

You mean you select the folder and which screen these needs to be added to?

> And also maybe a general batch edit option where multiple videos can be edited at once (similar to the mockup I posted on Feb 25th)

I'd like that feature as well.

Currently I thats on my roadmap (once im back onto Memebox Developing round about end of August):

0. Maybe I find some free time to do any of these features before

1. Finish the Release
- maybe if any of the above QoL features can be done "fast" I'll try to add em 
   -  in the memetime I'll create these features as `open for contribution` in the hope someone would like to try to create the initial work for that - any code helps I think :D
- there is still a lot to do for recipe's - which should also include a multi selection for "trigger random" 

2. Next Release: Memebox Twitch Extension 
- same as for the release I'll again try if those could be done "fast" 

3. Definitely QoL Stuff, maybe even between the Versions I'll try what I can - just lots todo on this project and i'm alone :D

**Comment by DrMinken** [comment]
First of, Love the work your doing here.  ❤️ 
Suggestion: 
1. Add remote support  - Make it possible to add other users/-mods to do add new/-edit and use Twitch Auth for login. Streamer can add by using the Twitch username to allow list. It's more to help the streamer to fix things on the fly.
2. Youtube - Don't know if this is possible to add in to video, but add a YT video by using the video code (exsampel : v=21X5lGlDOfg ) if you want to trigger something at a break time or wanna show of a video. Other great feature is add a viewer queue system with trigger !video v=21X5lGlDOfg and need to be approved before showing.

**Reply by negue** [reply]
Hey 👋 

sorry that it took so long to respond -.- and thank you! :) Been 2 years and I still haven't gotten it to a point where I want to be haha, some day I'll be awesome

To 1) That would be crazy awesome o.O - once the 2022.1 is released I wanted to start to work on a Twitch Extension for Memebox - for that I'll need a Hub (online Server) so that the events from the extension will be pushed to the running memebox instance. And once that exist I think this could be used for such a feature set - its just a ton of dependencies that have to be built first before something like this could be possible

To 2) for this the same dependencies ( minus a few) are needed like with 1) to get the full pontential - BUT- I think this could be already built up with current possibilies of memebox ... well I hope - I'd need to try to it first xD

How'd would you like it to be handled?

- Chatter suggests a video
- a mod checks that video and then marks that one as approved (with a chat command)
- the streamer / mod then triggers it with a command (or some streamdeck action like "trigger next video")

If its like that - I think I could build up a script and widget that could be used for that
